config/heroes/stronghold.json

```diff
@@ -95,11 +95,12 @@
 		"specialty" : {
 			"bonuses" : {
 				"offence" : {
-					"subtype" : "skill.offence",
-					"type" : "SECONDARY_SKILL_PREMY",
+					"subtype" : 0,
```

Comment:
Why is subtype here not skill.offence? Or something more descriptive than an index?

Reply:
Because this is just means melee. If many modders want, I can try to add new enum and map to json parser (similar subtype is used for 3 bonuses, not only my new). 

Skill as subtype - is no way. This PR is created to get rid from skill subtypes everywhere.

Reply:
subtype 0 is not that friendly to moders. They will have to remember always what 0,1 etc means.

Reply:
And is inconsistent with other skills
```
					"subtype" : "resource.gold",
					"type" : "GENERATE_RESOURCE",
```

Reply:
so, I need to add a strong identifier to all numeric subtypes (there is a many, for example, magic schools and land/sea movement, not only damage).

Reply:
@krs0 your comment is valid, but this is something extremely common in our bonus system.
Right now most (if not all) damage-affecting bonuses use this logic ( 0 = any, 1 = melee, 2 = ranged)

Some sort of automatic validation of bonuses + reduction of numeric ID's is nice thing to have, but I think it should probably be done separately. Don't want already massive PR to grow even more.

Reply:
@rilian-la-te Better option for now is probably to remove `subtype = 0` from json where this is redundant info. Example - eagle eye.
Oh, and in this case (Melee-only offence bonus) use subtype 2 instead.
Most bonuses use 0 - all attacks, 1 - shot only, 2 - melee only

Reply:
@IvanSavenko, no, GENERAL_DAMAGE_REDUCTION uses 0 for melee and 1 for ranged, my new bonus uses exactly same values. And I added 2 for GENERAL_DAMAGE_REDUCTION to use for armorer (e.g all). Can you show me an example where 0=any, 1=melee, 2=ranged is used?

Reply:
@IvanSavenko maybe leave PERCENTAGE_DAMAGE_BOOST consistent with GENERAL_DAMAGE_REDUCTION?

Reply:
Yeah, my bad. 
`SPELL_AFTER_ATTACK` and `SPELL_BEFORE_ATTACK` use 0/1/2, but damage-related use -1/0/1 (with -1 = any) or 0/1 (like you did)

Still, yet another example that bonus subID's need cleanup & unification...
(but not in this PR)

Reply:
@IvanSavenko so, I need to use -1 for armorer? understood. Will do. About string identifiers - agree, not for this PR.

Reply:
@rilian-la-te 
Armorer - either don't use subtype at all and use selector with only bonus type, or, if bonus supports different subtypes - use -1 = all, 0 = melee, 1 = ranged

Reply:
Bonus actually supports) subtype 0 is Shield spell, subtype 1 is Air Shield spell) so, -1)

Reply:
There are IDs for many object subtypes, maybe we could add more for some use caes?
https://wiki.vcmi.eu/Bonus_Format

Reply:
@DjWarmonger 
Yeah. Frankly, I'd even go further and convert *all* subID's to strings. Probably even use "spellLevel.3" form for spell levels (and other similar subtypes).
And also add validation for bonus parameters - make sure that bonus contains all fields used by said bonus, including addInfo and not just subtype.

But this is something for the future I think. Don't like too long & too large PR's.

Reply:
@IvanSavenko actually switched armorer to subtype -1 (it was used already for basilisk petrification, so, I just reused it).

Reply:
@IvanSavenko so, it seems than we need another bonus value type for handling such cases. Will look for it.

Reply:
@IvanSavenko it is ok to assume than bonuses from one source will become one defence factor? Or I need another assumption?

Reply:
or maybe assume than spell and non-spell bonuses should become separate defence factor?

Reply:
check DamageCalculator.cpp for details

In long-term I think *every* different base value from every bonus should be considered separately for damage calculation.

As of right now, calculator assumes that each different damage reduction source comes with a separate bonus. Armorer + Petrification work find right now simply because they use different bonuses.

So if some mod reuses `GENERAL_DAMAGE_REDUCTION` it might work not as intended.

Reply:
@IvanSavenko yes, and there is notes in json. I will do quick way - differentiate spell and non-spell effects, so, any non-spell effect will be additive to armorer, and any spell effect - to petrification. It will allow us to match h3 behaviour, but it will not fix unstackable shield spell.

Reply:
@IvanSavenko done by differentiating spell effect and not spell effect (AFAIK, basilisk bonus is SPELL_EFFECT one). But we really need to rework calculation to be selector-based at least for first 4 value type except additive).